TSDB: Implement downsampling on time-series indices

csoulios opened the pull request:
This PR implements downsampling operation on time series indices.

The PR creates a `_rollup` endpoint that allows users to downsample an index and can be
accessed by the following call:

```
POST /<source_index>/_rollup/<rollup_index>
{
    "fixed_interval": "1d"
}
```

### Requirements
An index can be downsampled if all of the following requirements are met:
 
 - Must be a time series index (have the `index.mode: time_series` index setting)
 - Must not be writeable (have the `index.blocks.write: true` index setting)
 - Must have dimension fields marked with mapping parameter `time_series_dimension: true`
 - Must have metric fields marked with mapping parameter `time_series_metric`


### Steps

The steps followed by the downsampling action are the following: 

1. Read the index mapping from the source index and extract the dimension and metric fields.
1. Create a temporary rollup index with the following settings/mapping:
    1. The following index settings will be copied from the original index
        - `index.mode`
        - `index.routing_path`
        - `index.time_series.start_time`
        - `index.time_series.end_time`
    1. Set the following index settings to the values of the original index:
        - `index.rollup.source.uuid`
        - `index.rollup.source.name`
    1. All fields mapped as `time_series_dimension` will be created in the rollup index with exactly the same mapping.
    1. All fields mapped as `time_series_metric` will be created and their field type will change to `aggregate_metric_double`.
1. Use the [`TimeSeriesIndexSearcher`](https://github.com/elastic/elasticsearch/blob/a89d4c34cbd8e5b3b3879df3aec632e7494e3c35/server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java) implementation to traverse over the (sorted by `_tsid`, `@timestamp`) source index and create the rollup index. 
1. Rename the temporary rollup index to its defined name.
1. If the source index is a member of a data stream, we must replace the original index with the rollup index. So the rollup index is added and the source index is deleted. The addition and deletion of the indices should be an atomic action.

### TODO 

- Implement logic for storing  fields that are neither dimensions nor metrics (aka labels)
- Make rollup task cancellable
- More testing

Relates to https://github.com/elastic/elasticsearch/issues/74660

Fixes #65769
Fixes #69799
Finally, this PR is based on the code written for https://github.com/elastic/elasticsearch/pull/64900

csoulios committed:
Remove feature flag es.rollup_v2_feature_flag_enabled

csoulios committed:
RollupShardIndexer

csoulios committed:
Cleanup

csoulios committed:
Test should not merge segments

csoulios committed:
Refactoring code

csoulios committed:
Fix wrong timestamp order

csoulios committed:
Cleanup

csoulios committed:
Cleanup

csoulios committed:
Removed CompressingOfflineSorter implementation

Not needed for timeseries indices

csoulios committed:
class to record

csoulios committed:
Added more tests

csoulios committed:
Refactored rollup action config

Removed most of the configuration, which is extracted from
the index mapping.

Modified TransportRollupAction to extract the rollup config from
the field caps

csoulios committed:
Refactored rollup action config

Removed most of the configuration, which is extracted from
the index mapping.

Modified TransportRollupAction to extract the rollup config from
the field caps.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @csoulios, I've created a changelog YAML for you.

csoulios committed:
Update docs/changelog/85708.yaml

csoulios committed:
Version updates

csoulios committed:
CI Fixes

csoulios committed:
checkstyle

csoulios committed:
checkstyle

csoulios committed:
Skip some failing tests

I will deal with them later

csoulios committed:
Fix broken test with wrong index sort order

csoulios committed:
Fix ILM test

csoulios commented on server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java:
```diff
@@ -78,7 +78,7 @@ public void search(Query query, BucketCollector bucketCollector) throws IOExcept
             @Override
             protected boolean lessThan(LeafWalker a, LeafWalker b) {
-                return a.timestamp < b.timestamp;
+                return a.timestamp > b.timestamp;
```
I have changed the sorting order because I was getting results in a wrong timestamp order when receiving documents from different segments. By changing this, the doc order problem was fixed. 

I think this bug is fixed by this PR: https://github.com/elastic/elasticsearch/pull/85526

csoulios replied:
Issue fixed by PR https://github.com/elastic/elasticsearch/pull/85526

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackClientPlugin.java:
```diff
@@ -411,12 +413,6 @@
         );
 
-        // rollupV2
-        if (RollupV2.isEnabled()) {
```
I removed the feature flag we had used for Rollups v2. Now, rollups code works as an operation for time_series indices

csoulios replied:
Added the TSDB feature flag

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupActionConfig.java:
```diff
@@ -6,157 +6,206 @@
+            ObjectParser.ValueType.STRING
         );
-        PARSER.declareObjectArray(optionalConstructorArg(), (p, c) -> MetricConfig.fromXContent(p), new ParseField(MetricConfig.NAME));
+        PARSER.declareStringOrNull(optionalConstructorArg(), new ParseField(TIME_ZONE));
```
Since the team has decided that we are going to support only the UTC timezone for now, maybe we should remove the `timezone` parameter from the `_rollup` operation and add it when we add support for other timezones.

csoulios replied:
The team decided to remove the `time_zone` parameter from the `_rollup` API and add it later more timezones are supported.

csoulios replied:
Removed `time_zone` rollup param at [d905038](https://github.com/elastic/elasticsearch/pull/85708/commits/d905038b8eb281c2933d209af57d4bd1e9a95d20)

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Enable test

csoulios committed:
Update docs/changelog/85708.yaml

elasticsearchmachine commented:
Hi @csoulios, I've updated the changelog YAML for you.

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Pull timestamp from AggregationExecutionContext

csoulios committed:
Remove time_zone param from rollup config

csoulios committed:
Fix build failure

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Added validations for rollup request parameters

Clean up handling of the temporary index name
Fixes #65769

csoulios committed:
Update docs/changelog/85708.yaml

elasticsearchmachine commented:
Hi @csoulios, I've updated the changelog YAML for you.

csoulios committed:
Set the number of shards and replicas

of the source index to the rollup index

csoulios committed:
Validate that index is read-only before rolling up

csoulios committed:
Delete source index after it has been downsampled

csoulios committed:
Added test and fix for rolling up data stream idx

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -250,112 +298,165 @@ public void onFailure(Exception e) {
+                if (sourceIndex.getParentDataStream() != null) {
+                    IndexMetadata rollupIndexMetadata = currentState.getMetadata().index(rollupIndexName);
+                    Index rollupIndex = rollupIndexMetadata.getIndex();
                     // If rolling up a backing index of a data stream, add rolled up index to backing data stream
-                    DataStream originalDataStream = originalIndex.getParentDataStream().getDataStream();
-                    List<Index> backingIndices = new ArrayList<>(originalDataStream.getIndices().size() + 1);
-                    // Adding rollup indices to the beginning of the list will prevent rollup indices from ever being
+                    DataStream originalDataStream = sourceIndex.getParentDataStream().getDataStream();
+                    List<Index> backingIndices = new ArrayList<>(originalDataStream.getIndices().size());
+                    // Adding the rollup index to the beginning of the list will prevent it from ever being
                     // considered a write index
                     backingIndices.add(rollupIndex);
-                    backingIndices.addAll(originalDataStream.getIndices());
+                    // Add all indices except the source index
+                    backingIndices.addAll(originalDataStream.getIndices().stream().filter(i -> i.getName() != sourceIndexName).toList());
                     DataStream dataStream = new DataStream(
                         originalDataStream.getName(),
                         backingIndices,
```
Adding the rollup index to the data stream and remove the source index, hopefully in an atomic action.

@martijnvg your feedback on this would be greatly appreciated. Thanks

martijnvg replied:
The data stream in cluster state is updated in a atomic way here, since
the original index is removed and rolled up index is added in the same `ClusterStateUpdateTask` instance.

csoulios replied:
Thank you. I pushed new code that uses `DataStream#replaceBackingIndex(...)`

csoulios committed:
minor change

csoulios committed:
More changes

csoulios committed:
Set max as default_metric for gauge metrics

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Minor change to address reviewer comments

csoulios committed:
Tidy up code for data streams

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupIndexerAction.java:
```diff
@@ -85,26 +115,28 @@ public ActionRequestValidationException validate() {
         @Override
         public int hashCode() {
-            return Objects.hash(rollupRequest);
+            return Objects.hash(rollupIndex, rollupRequest, dimensionFields, metricFields);
```
Good catch. Changed it to a more conventional `hashCode()` implementation.

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupActionConfig.java:
```diff
@@ -6,157 +6,205 @@
-            m.validateMappings(fieldCapsResponse, validationException);
-        }
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
```
Fixed

csoulios committed:
Changed hashCode

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -109,137 +109,184 @@ protected void masterOperation(
+                new ElasticsearchException(
+                    "Rollup requires setting [" + IndexMetadata.SETTING_BLOCKS_WRITE + " = true] for index [" + sourceIndexName + "]"
+                )
+            );
```
missing return statement?

csoulios replied:
Fixed in https://github.com/elastic/elasticsearch/pull/85708/commits/cd23fd713ba6adbaa3be8c3987f27900b6a0b476

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -109,137 +110,184 @@ protected void masterOperation(
+        if (state.getMetadata().index(tmpIndexName) != null) {
+            listener.onFailure(new ResourceAlreadyExistsException("Temporary rollup index [{}] already exists.", tmpIndexName));
             return;
         }
```
Without knowing all the details here, if this action is doing many things that are not atomic then there is room for failure here that needs to be taken into account. But maybe this action should only do the actual rollup and an the rollup ilm action should be changed to do the action listed below here? In case of failure, ILM can retry from the step that has failed.

nik9000 replied:
Should we remove the temporary index?

csoulios replied:
Removed temp index in [286cdbd](https://github.com/elastic/elasticsearch/pull/85708/commits/286cdbd3af0dcd6c504285d2f35be661a175ecd0)

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Return statement after error:

csoulios committed:
Merge branch 'master' into tsdb-rollup

imotov commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,137 +108,185 @@ protected void masterOperation(
                         if (indexerResp.isCreated()) {
-                            // 4.
+                            // 4. Make temp index read-only
+                            UpdateSettingsRequest updateSettingsReq = new UpdateSettingsRequest(WRITE_BLOCKED_SETTINGS, tmpIndexName);
```
I wonder if we should propagate parent task here as well.

csoulios replied:
Fixed. Now it propagates parent task for all client calls

imotov reviewed (commented):
Looks reasonable to me, but I would defer to @nik9000 for the final review.

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Fix compilation error

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Fix broken test

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Compute last_value for metric fields

salvatore-campagna commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/MetricFieldProducer.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    static class Max extends Metric {
+        private Double max;
```
I see we are using double here to keep track of metrics. In the aggregation framework using double as the underlying common data type is a source of issues/pain. Should we consider using BigDecimal? Precision would benefit...

csoulios replied:
You are right that double is potentially a source of drouble. However, all our aggs already use doubles everywhere. Also, `aggregate_metric_double` will store the result as a double

salvatore-campagna commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/MetricFieldProducer.java:
```diff
@@ -0,0 +1,237 @@
+                    metricValues.put(metric.name, metric.get());
+                }
+            }
+            return metricValues;
```
Maybe return as unmodifiable?

csoulios replied:
Fixed [ab1a4d8](https://github.com/elastic/elasticsearch/pull/85708/commits/ab1a4d899e60009b41a5c41b9b96f0c069723941)

nik9000 commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java:
```diff
@@ -320,15 +319,11 @@ private static List<NamedXContentRegistry.Entry> xContentEntries() {
                 ),
-                new NamedXContentRegistry.Entry(LifecycleAction.class, new ParseField(MigrateAction.NAME), MigrateAction::parse)
+                new NamedXContentRegistry.Entry(LifecycleAction.class, new ParseField(MigrateAction.NAME), MigrateAction::parse),
+                new NamedXContentRegistry.Entry(LifecycleAction.class, new ParseField(RollupILMAction.NAME), RollupILMAction::parse)
```
Does this need to live behind the tsdb feature flag?

csoulios replied:
Fixed in [86c3347](https://github.com/elastic/elasticsearch/pull/85708/commits/86c3347410aaf35140a79d839ff0494d1edcae35)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:
```diff
@@ -248,319 +184,212 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
+                        if (leafField.advanceExact(docId)) {
+                            for (int i = 0; i < leafField.docValueCount(); i++) {
+                                Object obj = leafField.nextValue();
+                                if (obj instanceof Number number) {
```
I talked with @csoulios and it's something that he does want to move out but would like to save for a follow up change.

csoulios committed:
Removed temporary rollup index

Rollups will be stored in the rollup index directly

csoulios committed:
Refresh the rollup index in the end

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/MetricFieldProducer.java:
```diff
@@ -135,40 +131,25 @@ void reset() {
+                    case "max" -> list.add(new Max());
+                    case "sum" -> list.add(new Sum());
+                    case "value_count" -> list.add(new ValueCount());
+                    default -> throw new IllegalArgumentException("Unsupported metric type [" + metricName + "]");
```
I totally changed this part. Now there are two implementations of `MetricFieldProducer` (`GaugeMetricFieldProducer` and `CounterMetricFieldProducer`) that are created based on metric type of the field

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupActionConfig.java:
```diff
@@ -6,157 +6,205 @@
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeString(FIXED_INTERVAL);
+        fixedInterval.writeTo(out);
+        out.writeString(timeZone);
```
I chose to serialize timezone here because I thought we are going make it configurable in the future., so we better send a fixed string now. This will keep the protocol stable in the future. The same for interval type, I wanted to keep this information.

But you never know we may never find it useful to add support for multiple timezones and interval. We will definitely find something else to add there. So, for the sake of YAGNI principle, you think I should remove them?

csoulios committed:
Use randomValueOtherThan

csoulios commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -78,7 +78,7 @@ public class TimeseriesLifecycleType implements LifecycleType {
         MigrateAction.NAME,
         FreezeAction.NAME,
-        RollupV2.isEnabled() ? RollupILMAction.NAME : null
+        RollupILMAction.NAME
```
The rollup action will be available but it will not operate on indices that are not time series indices. 

So, in a similar way as we did for `TimeSeriesIdFieldMapper` I did not put this behind a feature flag. I can add it though, if you think it makes sense.

csoulios replied:
Fixed in [86c3347](https://github.com/elastic/elasticsearch/pull/85708/commits/86c3347410aaf35140a79d839ff0494d1edcae35)

csoulios commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/RollupILMActionTests.java:
```diff
@@ -82,9 +81,11 @@ RollupILMAction notCopy(RollupILMAction rollupILMAction) {
-                List<MetricConfig> metricConfigs = new ArrayList<>(rollupILMAction.config().getMetricsConfig());
-                metricConfigs.add(new MetricConfig(randomAlphaOfLength(4), Collections.singletonList("max")));
-                newConfig = new RollupActionConfig(rollupILMAction.config().getGroupConfig(), metricConfigs);
+                DateHistogramInterval fixedInterval = ConfigTestHelpers.randomInterval();
```
I didn't know there was such a method. Thanks :)

Done in [c22b6ba](https://github.com/elastic/elasticsearch/pull/85708/commits/c22b6baadf30b95880e0218e195270be3ffba5be)

csoulios committed:
Do not serialize constants in RollupActionConfig

csoulios committed:
Add tsdb feature flag

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/Rollup.java:
```diff
@@ -160,15 +156,13 @@ public List<RestHandler> getRestHandlers(
-                new ActionHandler<>(XPackInfoFeatureAction.ROLLUP, RollupInfoTransportAction.class)
+                new ActionHandler<>(XPackInfoFeatureAction.ROLLUP, RollupInfoTransportAction.class),
+                // Rollup / Downsampling
+                new ActionHandler<>(RollupIndexerAction.INSTANCE, TransportRollupIndexerAction.class),
```
Fixed in [86c3347](https://github.com/elastic/elasticsearch/pull/85708/commits/86c3347410aaf35140a79d839ff0494d1edcae35)

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
minor change

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/MetricFieldProducer.java:
```diff
@@ -0,0 +1,237 @@
+import java.util.List;
+import java.util.Map;
+
+abstract class MetricFieldProducer {
```
It's probably worth sprinkling javadoc in a bunch of places, here included.

csoulios replied:
Fixed in [de6b766](https://github.com/elastic/elasticsearch/pull/85708/commits/de6b766575f2ab5cd64067d6dced334d14c07208)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:
```diff
@@ -229,14 +168,18 @@ public void afterBulk(long executionId, BulkRequest request, BulkResponse respon
-                numSent.addAndGet(-items);
+                if (failure != null) {
+                    long items = request.numberOfActions();
+                    numSent.addAndGet(-items);
```
Even if it failed it was still "sent", right?

csoulios replied:
I think here there may be a problem with retries. If failures are retried, then same docs are counted multiple times numSent. Right?

In the end we require that numSent == numIndexed, so if there are retries we may get failures because of this

csoulios replied:
Fixed in [839a29d](https://github.com/elastic/elasticsearch/pull/85708/commits/839a29df872cfdae88af8c8885297e8811ced7b3)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:
```diff
@@ -229,14 +168,18 @@ public void afterBulk(long executionId, BulkRequest request, BulkResponse respon
                         );
-                    logger.error("failures: [{}]", failures);
+                    numFailed.addAndGet(failures.size());
+                    logger.error("Shard {} failed to populate rollup index: [{}]", indexShard.shardId(), failures);
```
Should this abort the load process?

nik9000 replied:
We're holding off on early abort for a follow up PR.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:
```diff
@@ -229,14 +168,18 @@ public void afterBulk(long executionId, BulkRequest request, BulkResponse respon
+                    long items = request.numberOfActions();
+                    numSent.addAndGet(-items);
+                    numFailed.addAndGet(items);
+                }
```
Should this abort the load process?

nik9000 replied:
We're holding off on early abort for a follow up PR.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:
```diff
@@ -248,319 +191,203 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
+    private class TimeSeriesBucketCollector extends BucketCollector {
+        private final BulkProcessor bulkProcessor;
+        private long docsProcessed;
+        private long bucketsCreated;
```
I think it'd be useful to plumb these into the task so the get task API can show you the progress. It'd be fine to do in a follow up though.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -277,126 +398,126 @@ private static XContentBuilder getDynamicTemplates(XContentBuilder builder) thro
+            @Override
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                // 7. Delete the source index
+                deleteSourceIndex(sourceIndexName, rollupIndexName, listener);
```
Will we always want to delete the source index? I get that we do if we're replacing the source in the data stream but if it isn't in a data stream is that required? Maybe I'm thinking about this funny - maybe we have to because of index patterns or something.

csoulios replied:
Do not deleting the source index when an index is not a data stream backing index has certain benefits (mainly flexibility), but also adds complexity in certain areas such as ILM.

We have discussed this topic in the past and we have agreed that the souce index should be deleted in any case, even if it's not a data stream index.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -277,126 +398,126 @@ private static XContentBuilder getDynamicTemplates(XContentBuilder builder) thro
+                if (acknowledgedResponse.isAcknowledged()) {
+                    // Source index was deleted successfully.
+                    // 8. Finalize the rollup index by setting its state to SUCCESS
+                    finalizeRollupIndex(rollupIndex, listener);
```
Is it possible to submit the delete action and for the source index to be deleted and then something to go wrong and you don't get the response here so the rollup index isn't "finalized"? I think it'd be nice to make the replacement in the data stream, the delete, and the finalize atomic in one action.

csoulios replied:
I modified the code so that the following operations happen within a single state update request, so they are all performed atomically.

 - If source index belongs to a data stream, add the rollup index and remove the source index
 - Mark rollup index as completed successfully
 - Delete the source index


See [ee42ef5](https://github.com/elastic/elasticsearch/pull/85708/commits/ee42ef5aacc5328ceef64b4f052f2e11a0537f74)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,137 +104,161 @@ protected void masterOperation(
+            ).mappings(XContentHelper.convertToJson(BytesReference.bytes(mapping), false, XContentType.JSON));
+
+            // 2. Create rollup index
+            submitUnbatchedTask("create-rollup-index", new ClusterStateUpdateTask() {
```
I think maybe we can't use `submitUnbatchedTask`, right? The docs say " This method exists for legacy reasons but is deprecated and forbidden in new production code because unbatched tasks are a source of performance and stability bugs."

csoulios replied:
I have seen the deprecation comment and researched a bit more into this. Method `clusterService.submitUnbatchedStateUpdateTask()` is deprecated by currently used all over the place. 

My goal was to get this PR merged and then work in this issue in a separate PR. 

@DaveCTurner do you have an opinion if we should merge this PR while using the deprecated method?

DaveCTurner replied:
I would very much prefer to avoid adding more unbatched tasks, they are truly awful for performance in large clusters. Here for instance if ILM were paused for a while then it might need to create thousands of rollup indices when restarted, and as currently written it would put all of those tasks in the queue at once and then work through them one-at-a-time.

That said, here it looks like we're just reworking an existing unbatched task rather than adding a new one right? If so this isn't really changing anything in terms of the cluster state tasks so it seems ok to proceed.

DaveCTurner replied:
> That said, here it looks like we're just reworking an existing unbatched task rather than adding a new one right? 

Actually now that I look closer it seems a bit more than that, we're introducing tasks that weren't there before which makes me concerned. Is there a good reason to avoid batching these these things? It shouldn't be too hard. See `MetadataIndexTemplateService#TEMPLATE_TASK_EXECUTOR` for a reasonably simple example.

csoulios replied:
@DaveCTurner thanks for the feedback.

csoulios replied:
Fixed in [332bef6](https://github.com/elastic/elasticsearch/pull/85708/commits/332bef648bbbddae58756acc40fca1adc9d3b257)

nik9000 commented on x-pack/plugin/security/qa/operator-privileges-tests/build.gradle:
```diff
@@ -39,7 +39,6 @@ testClusters.configureEach {
   setting 'path.repo', repoDir.absolutePath
 
   requiresFeature 'es.user_profile_feature_flag_enabled', Version.fromString("8.1.0")
-  requiresFeature 'es.rollup_v2_feature_flag_enabled', Version.fromString("8.0.0")
```
Should this have the time series flag instead?

csoulios replied:
Fixed in [270ea69](https://github.com/elastic/elasticsearch/pull/85708/commits/270ea69604e30fc6f3e1fc6993f9ddf8d84c91ba)

csoulios committed:
Perform the following actions in one state update:

- Replace source with rollup in the datastream
- Delete source index
- Mark rollup index as status=success

The above actions will be performed in a single atomic step

csoulios committed:
Perform batched cluster state update

csoulios committed:
Cleanup the mess

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
More cleanup

csoulios committed:
Added feature flag to security tests

csoulios committed:
Added feature flag to build files

csoulios committed:
Reverted wrong version bump

csoulios committed:
Add the number of indexed docs to the shard response

csoulios committed:
Added javadoc

nik9000 reviewed (commented):
I looked at it for a while and left some comments. I think I want to have another look on Monday too.

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +129,165 @@ protected void masterOperation(
+        // 5. Add rollup index to the data stream
+        // 6. Mark rollup index as "completed successfully"
+        // 7. Delete the source index
+        // 8. Refresh rollup index
```
Happy to do it, but I am not sure I understand why data may vanish

nik9000 replied:
It's just vanish for a few seconds. If you put the new index into the data stream and we start querying it *before* an explicit refresh then some of the rollup that your wrote may not yet be visible. So data will appear to "vanish" until the refresh takes.

csoulios replied:
Fixed in [0ff80dc](https://github.com/elastic/elasticsearch/pull/85708/commits/0ff80dc4792a1caa76bbb44566e9f4f6e5c4774d)

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +129,165 @@ protected void masterOperation(
+                } else if (e.getValue().values().iterator().next().getMetricType() != null) {
+                    metricFieldCaps.put(field, fieldCaps);
+                } else {
+                    // TODO: Field is not a dimension or a metric. Treat it as a tag
```
Yes, in a following PR we will add the tags. I left this there as a placeholder

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +129,165 @@ protected void masterOperation(
-                                        if (resizeResponse.isAcknowledged()) {
-                                            // 6.
-                                            publishMetadata(originalIndexName, tmpIndexName, rollupIndexName, listener);
+                                    // 5. Add rollup index to data stream and publish rollup metadata
```
Not sure why we should wait for the replicas to recover. I was thinking that this can happen in parallel. Any concerns we may lose the primary shard while deleting the source index?

I am afraid that waiting for replicas to recover may take lots of time.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -277,132 +421,157 @@ private static XContentBuilder getDynamicTemplates(XContentBuilder builder) thro
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, sourceIndexMetadata.getNumberOfShards())
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(IndexMetadata.INDEX_ROLLUP_STATUS.getKey(), IndexMetadata.RollupTaskStatus.STARTED)
+                .build()
```
Maybe we should set the refresh_interval to -1 here. The default interval tries to make data visible while writing at the cost of less efficient writes.

nik9000 replied:
Also - I wonder if we should force_merge after we're done writing. It doesn't have to be now, but it should make the target index more efficient to query and we're not going to write to it again.

csoulios replied:
Fixed in [0ff80dc](https://github.com/elastic/elasticsearch/pull/85708/commits/0ff80dc4792a1caa76bbb44566e9f4f6e5c4774d)

- Set refresh_interval to -1
- Force merge index in the end

weizijun commented:
I added a PR(https://github.com/elastic/elasticsearch/pull/82944) in January, and discuss with @csoulios about the downsampling design several times. Now, I'm waiting for this PR to be merged. Then I can contribute my code together to the downsampling requirement.

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Changes to rollup index:

- Set refresh_interval to -1
- Refresh index before adding to data stream
- Force merge index in the end

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -277,132 +421,157 @@ private static XContentBuilder getDynamicTemplates(XContentBuilder builder) thro
+        ).settings(
+            /*
+             * When creating the rollup index, we copy the index.number_of_shards from source index,
+             * and we set the index.number_of_replicas to 0, to avoid replicating the temp index.
```
Fixed in [0ff80dc](https://github.com/elastic/elasticsearch/pull/85708/commits/0ff80dc4792a1caa76bbb44566e9f4f6e5c4774d)

csoulios committed:
Remove org.elasticsearch.rollup from module-info

csoulios committed:
Merge branch 'master' into tsdb-rollup

csoulios committed:
Added more validation for source index

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +129,165 @@ protected void masterOperation(
+             */
+            for (Map.Entry<String, Map<String, FieldCapabilities>> e : fieldCapsResponse.get().entrySet()) {
+                String field = e.getKey();
+                FieldCapabilities fieldCaps = e.getValue().values().iterator().next();
```
Fixed in [1dd28fd](https://github.com/elastic/elasticsearch/pull/85708/commits/1dd28fdba12c1b9865ceb57e414001cc013da59a)

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +133,214 @@ protected void masterOperation(
+                                                    );
+                                                }
+                                            },
+                                                e -> deleteRollupIndex(
```
I wonder if there is a good way to reduce the duplicate calls here. Oh well, in a follow up.

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -108,144 +133,214 @@ protected void masterOperation(
+                                                                new ElasticsearchException(
+                                                                    "Failed to force-merge index [" + rollupIndexName + "]",
+                                                                    e
+                                                                )
```
Will this fail the request but not cause the destination index to be deleted? What will ILM think of that?

csoulios replied:
In [393d7c2](https://github.com/elastic/elasticsearch/pull/85708/commits/393d7c2cd2c8c75f88970ca6caee55c2be9b1f08), I changed it so that it does not fail the rollup action.

nik9000 reviewed (approved):
Let's get this in. There's certainly more to do but I feel like this is a good start for rollups.

csoulios committed:
Do not fail rollup if force-merge fails

csoulios committed:
Merge branch 'master' into tsdb-rollup

nik9000 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:
```diff
@@ -267,12 +273,19 @@ protected void masterOperation(
+                                                                 * operation.
+                                                                 */
+                                                                logger.error(
+                                                                    "Failed to force-merge rollup index [" + rollupIndexName + "]",
```
:+1:

It's probably worth a test that fails here and makes sure we don't fall over. I think this will do the job but it's probably worth tracing this path. Both as a "for now" thing and to make sure that changes in the future don't cause us to lose this property.

csoulios committed:
Merge branch 'master' into tsdb-rollup
